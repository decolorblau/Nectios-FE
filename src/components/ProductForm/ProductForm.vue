<template>
  <v-row justify="center" class="form-container">
    <v-col cols="10" sm="10" md="8" lg="6">
      <v-card ref="onSubmit" class="pa-3 product-card">
        <v-card-title class="product mt-12 mb-10">
          <h2 class="product__title">
            ADD YOUR PRODUCTS TO<span class="world"> BLUE<span class="blue">WORLD</span></span>
          </h2>
        </v-card-title>
        <v-card-text>
          <v-text-field
            ref="name"
            v-model="name"
            :rules="[() => !!name || 'This field is required']"
            label="Name"
            required
            color="#207eeb"
            placeholder="Campbell's"
          ></v-text-field>
          <v-text-field
            outlined
            ref="price"
            v-model="price"
            :rules="[() => !!price || 'This field is required']"
            label="Price"
            placeholder="2.50"
            type="number"
            color="#207eeb"
            required
          ></v-text-field>
          <v-textarea
            ref="description"
            rows="5"
            row-height="30"
            v-model="description"
            maxlength="400"
            :rules="[() => !!description || 'This field is required']"
            label="Description"
            counter="400"
            color="#207eeb"
            single-line
            required
            placeholder="Campbell's condensed is a tomato Soup "
          ></v-textarea>
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn
            variant="outlined"
            class="mb-1 mr-2"
            :disabled="name === '' || description === '' || price === ''"
            color="#207eeb"
            text
            @click="onSubmit"
          >
            CREATE PRODUCT
          </v-btn>
        </v-card-actions>
      </v-card>
      <div class="errorMessage">
        <v-alert v-if="wrongProduct" variant="outlined" density="compact" color="#207eeb">
          ! This is a wrong Product
        </v-alert>
      </div>
    </v-col>
  </v-row>
</template>

<script>
import { defineComponent } from "vue";
import { mapActions, mapGetters, mapState } from "vuex";

export default defineComponent({
  name: "ProductForm",
  data() {
    return {
      name: "",
      description: "",
      price: 0,
      isDisabled: true,
      wrongProduct: false,
    };
  },
  computed: {
    ...mapState(["user"]),
  },
  methods: {
    ...mapActions(["addProduct"]),
    ...mapGetters(["redirectToHome"]),
    async onSubmit() {
      try {
        this.wrongProduct = false;
        if (this.name !== "" && this.description !== "" && this.price !== "") {
          const { clientKey } = JSON.parse(
            // eslint-disable-next-line comma-dangle
            localStorage.getItem("clientKey") || ""
          );

          const productData = {
            name: this.name,
            description: this.description,
            price: this.price,
            clientKey,
          };
          const status = await this.addProduct(productData);
          if (status === 200) {
            this.redirectToHome();
          } else {
            this.wrongProduct = true;
          }
        }
      } catch {
        this.wrongProduct = true;
      }
    },
  },
});
</script>

<style lang="scss" scoped>
.product {
  text-align: center;
  align-items: center;
  justify-content: center;
  &__title {
    font-size: 26px;
    font-family: "Roboto, sans-serif";
    font-weight: 400;
  }
}
.form-container {
  width: 100%;
  margin-top: 30px;
  padding: 0;
}
.product-card {
  box-shadow: 0 0 10px rgba(32, 126, 235, 0.2);
  font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
}
.errorMessage {
  margin-top: 20px;
  height: 60px;
  display: flex;
  align-items: center;
  justify-content: center;
}
.blue {
  color: #207eeb;
}
.world {
  font-weight: 600;
}
</style>
