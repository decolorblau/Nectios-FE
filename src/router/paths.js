const paths = {
  home: "/",
  detail: "/products/:id",
  login: "/login",
  newProduct: "/new",
  user: "/user",
  notFound: "/:pathMatch(.*)*",
};

export default paths;
