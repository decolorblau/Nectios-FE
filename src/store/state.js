const state = {
  products: [],
  comments: [],
  user: {
    isAuthenticated: false,
    clientKey: "",
    user: {
      email: "",
      name: "",
      profilePhoto: "",
      role: "",
      surname: "",
    },
  },
  currentProduct: {
    product: {
      id: "",
      client_id: "",
      description: "",
      key: "",
      name: "",
      price: "",
      product_photo: "",
    },
    comments: [],
  },
};

export default state;
