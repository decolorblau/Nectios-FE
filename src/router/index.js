import { createRouter, createWebHistory } from "vue-router";
import Home from "../views/Home.vue";
import Login from "../views/LoginView.vue";
import NotFound from "../views/NotFoundView.vue";
import paths from "./paths";
import User from "../views/UserView.vue";
import Detail from "../views/DetailView.vue";

// eslint-disable-next-line object-curly-newline
const { home, notFound, login, user, detail } = paths;

const routes = [
  {
    path: home,
    name: "Home",
    component: Home,
  },
  {
    path: login,
    name: "Login",
    component: Login,
  },
  {
    path: user,
    name: "User",
    component: User,
  },
  {
    path: detail,
    name: "Detail",
    component: Detail,
  },
  {
    path: notFound,
    name: "NotFound",
    component: NotFound,
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

export default router;
