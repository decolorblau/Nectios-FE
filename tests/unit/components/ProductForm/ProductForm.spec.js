import ProductForm from "@/components/ProductForm/ProductForm.vue";
import { shallowMount, mount } from "@vue/test-utils";
import { createStore } from "vuex";
import router from "@/router";
import state from "../../mockState";

describe("Given a ProductForm component", () => {
  describe("When is rendered", () => {
    test("Then it matches de snapshot", () => {
      const wrapper = shallowMount(ProductForm);
      expect(wrapper.html()).toMatchSnapshot();
    });
    test("Then it should render h2 tag with ADD YOUR PRODUCTS TO BLUEWORLD text", () => {
      const wrapper = shallowMount(ProductForm);
      const h2Value = wrapper.find("h2").text();
      expect(h2Value).toBe("ADD YOUR PRODUCTS TO BLUEWORLD");
    });
    test("Then it should render Name input", () => {
      const wrapper = shallowMount(ProductForm);
      const [nameInput] = wrapper.findAll("v-text-field");
      expect(nameInput.attributes("label")).toBe("Name");
    });
    test("Then it should render Description input", () => {
      const wrapper = shallowMount(ProductForm);
      const [, priceInput] = wrapper.findAll("v-text-field");
      expect(priceInput.attributes("label")).toBe("Price");
    });
    test("Then it should render Description input", () => {
      const wrapper = shallowMount(ProductForm);
      const descriptionInput = wrapper.find("v-textarea").attributes("label");
      expect(descriptionInput).toBe("Description");
    });
    test("Then it should render Submit button", () => {
      const wrapper = shallowMount(ProductForm);
      const submitButton = wrapper.find("v-btn").text();
      expect(submitButton).toBe("CREATE PRODUCT");
    });
    describe("When the form is submitted", () => {
      test("Then it should invoke onSubmit", async () => {
        const store = createStore({
          state() {
            return state;
          },
          actions: {
            addProduct: jest.fn(),
          },
        });

        const wrapper = mount(ProductForm, {
          global: {
            plugins: [router, store],
          },
        });

        const onSubmit = jest.fn();
        onSubmit();
        const form = wrapper.get("v-card");
        await form.trigger("submit");
        expect(onSubmit).toHaveBeenCalled();
      });
    });
  });
});
