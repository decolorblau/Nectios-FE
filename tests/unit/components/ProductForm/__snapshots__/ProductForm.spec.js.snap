// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ProductForm component When is rendered Then it matches de snapshot 1`] = `
<v-row justify="center" class="form-container">
  <v-col cols="10" sm="10" md="8" lg="6">
    <v-card class="pa-3 product-card">
      <v-card-title class="product mt-12 mb-10">
        <h2 class="product__title"> ADD YOUR PRODUCTS TO<span class="world"> BLUE<span class="blue">WORLD</span></span></h2>
      </v-card-title>
      <v-card-text>
        <v-text-field modelvalue="" rules="() => {
                      /* istanbul ignore next */
                      cov_1rnru8588v.f[9]++;
                      cov_1rnru8588v.s[43]++;
                      return (
                        /* istanbul ignore next */
                        (cov_1rnru8588v.b[5][0]++, !!_ctx.name) ||
                        /* istanbul ignore next */
                        (cov_1rnru8588v.b[5][1]++, 'This field is required')
                      );
                    }" label="Name" required="" color="#207eeb" placeholder="Campbell's"></v-text-field>
        <v-text-field outlined="" modelvalue="0" rules="() => {
                      /* istanbul ignore next */
                      cov_1rnru8588v.f[11]++;
                      cov_1rnru8588v.s[45]++;
                      return (
                        /* istanbul ignore next */
                        (cov_1rnru8588v.b[7][0]++, !!_ctx.price) ||
                        /* istanbul ignore next */
                        (cov_1rnru8588v.b[7][1]++, 'This field is required')
                      );
                    }" label="Price" placeholder="2.50" type="number" color="#207eeb" required=""></v-text-field>
        <v-textarea rows="5" row-height="30" modelvalue="" maxlength="400" rules="() => {
                      /* istanbul ignore next */
                      cov_1rnru8588v.f[13]++;
                      cov_1rnru8588v.s[47]++;
                      return (
                        /* istanbul ignore next */
                        (cov_1rnru8588v.b[9][0]++, !!_ctx.description) ||
                        /* istanbul ignore next */
                        (cov_1rnru8588v.b[9][1]++, 'This field is required')
                      );
                    }" label="Description" counter="400" color="#207eeb" single-line="" required="" placeholder="Campbell's condensed is a tomato Soup "></v-textarea>
      </v-card-text>
      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn variant="outlined" class="mb-1 mr-2" disabled="true" color="#207eeb" text=""> CREATE PRODUCT </v-btn>
      </v-card-actions>
    </v-card>
    <div class="errorMessage">
      <!--v-if-->
    </div>
  </v-col>
</v-row>
`;
